src/log.py

```diff
@@ -0,0 +1,57 @@
+import logging
+from pythonjsonlogger import jsonlogger
+import dateutils
+
+def setup_root_logger():
+    root_logger = logging.getLogger("")
+
+    # output to stderr
+    _handler = logging.StreamHandler()
+    _handler.setLevel(logging.INFO)
+
+    class FormatterWithEncodedExtras(logging.Formatter):
+        def format(self, record):
+            # exclude all known keys in Record
+            # bundle the remainder into an 'extra' field,
+            # bypassing attempt to make Record read-only
+            _known_keys = [
+                'asctime', 'created', 'filename', 'funcName', 'levelname', 'levelno', 'lineno',
+                'module', 'msecs', 'message', 'name', 'pathname', 'process', 'processName',
+                'relativeCreated', 'thread', 'threadName',
+                # non-formatting fields present in __dict__
+                'exc_text', 'exc_info', 'msg', 'args',
+            ]
+            unknown_fields = {key: val for key, val in record.__dict__.items() if key not in _known_keys}
+            record.__dict__['extra'] = dateutils.json_dumps(unknown_fields)
+            return super(FormatterWithEncodedExtras, self).format(record)
+
+    _handler.setFormatter(FormatterWithEncodedExtras('%(levelname)s - %(asctime)s - %(message)s -- %(extra)s'))
+
+    root_logger.addHandler(_handler)
+    root_logger.setLevel(logging.DEBUG)
+
+def json_formatter():
+    _supported_keys = [
```

Comment:
Remove the unused items? also, the leading `_` in the variable names are redundant in this use case.

Reply:
the commented out items serve to show us what _can_ be included vs what _is_ being included.

the underscores denoting private/public scope are something I tend to avoid so I'm not 100% sure this was my code, but I've never applied it consistently. ultimately it doesn't really bother me.

Reply:
The underscores are there because this code comes from the top level of a module, but can now be removed